### PR TITLE
feat(upgrade): add upgrade framework in controllers

### PR DESCRIPTION
**Pull request by shubham14bajpai**
Signed-off-by: shubham <shubham.bajpai@mayadata.io>

This PR adds an upgrade framework for all the controllers to perform any upgrade steps from within the controller itself to avoid any pod exec issues.

### CODE REVIEW

File: pkg/controllers/cspc-controller/sync.go [hunk 1]
```diff
@@ -36,6 +38,17 @@ import (
 	"k8s.io/klog"
 )
 
+type upgradeParams struct {
+	cspc   *cstor.CStorPoolCluster
+	client clientset.Interface
+}
+
+type upgradeFunc func(u *upgradeParams) (*cstor.CStorPoolCluster, error)
```
Comment (mittachaitu) [inline pkg/controllers/cspc-controller/sync.go:46]: Will this make etcd calls to update CSPC? If yes just thinking why can't we make this internal Upgrade as an atomic operation? Is there any reason why we not selected?

File: pkg/controllers/cstorvolumeconfig/controller.go [hunk 2]
```diff
@@ -156,6 +169,25 @@ func (c *CVCController) syncCVC(cvc *apis.CStorVolumeConfig) error {
 	}
 
 	var err error
+
+	cvc, err = c.reconcileVersion(cvc)
```
Comment (mittachaitu) [inline pkg/controllers/cstorvolumeconfig/controller.go:173]: Before calling reconcileVersion, we need to populate CVC version details.

**Comment by shubham14bajpai** [comment]
> Can we use these functions not only to perform exec but also for any schema changes i.e adding field? Instead of prepopulating by upgrade job?

Yes we can edit the fields according to requirements in these functions also.

File: pkg/controllers/cspc-controller/sync.go (hunk 1 above)
Comment (shubham14bajpai) [reply pkg/controllers/cspc-controller/sync.go:46]: Some times the operations may not be atomic or we have steps that need to be performed in some order. If by atomic you mean single operation then it is safer to have a framework that allows you to have both options of having separate or a single operation. We can always write functions that don't make any etcd calls if not required but having the option will be better for any future requirements.

File: pkg/controllers/cstorvolumeconfig/controller.go (hunk 2 above)
Comment (shubham14bajpai) [reply pkg/controllers/cstorvolumeconfig/controller.go:173]: Version will be populated while creation itself for CVC.

File: pkg/controllers/cspc-controller/sync.go [hunk 3]
```diff
@@ -71,6 +77,32 @@ func (c *Controller) sync(cspc *cstor.CStorPoolCluster, cspiList *cstor.CStorPoo
...
+			err,
+		)
+		_, err = c.clientset.CstorV1().CStorPoolClusters(cspcGot.Namespace).Update(cspcGot)
+		if err != nil {
+			klog.Errorf("failed to update versionDetails status for cspc %s:%s", cspcGot.Name, err.Error())
+		}
+		return nil
+	}
+
+	if ok, reason := c.ShouldReconcile(*cspc); !ok {
```
Comment (mittachaitu) [inline pkg/controllers/cspc-controller/sync.go:98]: It should move below the code
```go
	if !cspcGot.DeletionTimestamp.IsZero() {
}
```

Comment (shubham14bajpai) [reply pkg/controllers/cspc-controller/sync.go:98]: When the controller is upgraded then reconcile version will never be triggered after patching the VersionDetails.DesiredVersion to the upgraded version as the ShouldReconcile function checks for CurrentVersion. Once the reconcileVersion is successful the CurrentVersion on the CR will be changed to the new version.

Comment (mittachaitu) [reply pkg/controllers/cspc-controller/sync.go:98]: I got it but we can delete the resource irrespective of current version so need to move below https://github.com/openebs/cstor-operators/blob/bc2114c8698a27a602fb55f419618de7e9fe16d4/pkg/controllers/cspc-controller/sync.go#L82

(Or)

[This snippet](https://github.com/openebs/cstor-operators/blob/bc2114c8698a27a602fb55f419618de7e9fe16d4/pkg/controllers/cspc-controller/sync.go#L82) should move before any function.

Comment (shubham14bajpai) [reply pkg/controllers/cspc-controller/sync.go:111]: Got it! Nice Catch :+1:

**Pull request merged**